Add_Keyboard_Support/InterruptHandling.md

```diff
@@ -0,0 +1,109 @@
+# Handling the keyboard interrupt
+
+Although we can use either PIC or IOAPIC to handle the Keyboard interrupt, for this section we will refer to the IOAPIC implementation, btw the only difference is how to configure the IRQ and send the EOI command, so if you are supporting the PIC most of the information in this guide will still be useful. 
+
+## IRQ and IO-Apic
+
+* The Keyboard IRQ is the number 1. This corresponds to pin 1 on the IO Apic, that is controlled by entry 1 in Redirection Table, 
+* The entry #1 of the redirection table is accessed as 2x32bit registers where the register number is defined as follows: a
+
+```
+redtbl_offset = 0x10 + (entry_number * 2)
+```
+
+in this case then we have the offset for our entry at:  12h and 13h (called IOREDTBL1), where 0x12 are the lower 32bits of the table entry. 
+
+Before unmasking the IRQ bit for the keyboard we need to implement a function (we can leave it empty for now) and set it into the IDT, for this section we will call the function `keyboard_irq_handler`:
+
+```c
+void keyboard_irq_handler() {
+
+}
+```
+ 
+Then once the function on the IDT vector has been set, we need to unmask the mask bit on the IRQ Redirection Table, and set the destination apic id (Refer to the APIC for details on the fields of a IO Redirection table)
+
+
+## Driver Information
+
+The ps2 keyboards uses two IO Ports for communication with the cpu: 
+
+| IO Port | Access Type | Purpose                                                         |
+|---------|-------------|-----------------------------------------------------------------|
+|  0x60   | R/W         | Data Port                                                       | 
+|  0x64   | R/W         | On read is the Status Register on write is the command register | 
+
+* There are three different sets of scancodes, is a good idea to check with the keyboard what scancode is being used. 
+* Usually the PS/2 controller (the one that the OS usually talks to using ports 0x60 and 0x64) convert set2 scancodes into set1 (for legacy reasons)
+* To check if the translation is enabled, the command 0x20 must be sent on port 0x64, and then read the byte on 0x60. If the 6th bit is set than the translation is enabled. 
+* If we want to disable the translation we need: 
+   - Read current controller configuration byte, with the 0x20 command (the byte will be sent on port 0x60)
+   - Clear the 6th on the current controller configuration byte
+   - Send the command 0x60 (it says that next byte on data port has to be written as the controller configuration byte) on port 0x64, then send the updated configuration byte to port 0x60
+   - For our driver we will keep the translation enabled
+* The only scancode set guaranted to be supported by keyboards is the set2. Keep in mind that most of the time the kernel communicate with the 8042_PS2 Controller, and in this case the scancodes can be translated into set1.
```

Comment:
the kernel will always talk to the ps2 controller, so you'll always have set2 to set1 translation available. 
The controller is responsible for all the electrical and sometimes mechanical stuff of the ps2 ports, stuff the kernel cant deal with.